src/style/style.js

```diff
@@ -307,25 +299,48 @@ class Style extends Evented {
         for (const sourceId in this.sourceCaches)
             this.sourceCaches[sourceId].used = false;
 
-        this._updateZoomHistory(z);
+        const parameters = {
+            zoom: z,
+            now: Date.now(),
+            defaultFadeDuration: 300,
+            zoomHistory: this._updateZoomHistory(z)
+        };
 
         for (const layerId of this._order) {
             const layer = this._layers[layerId];
 
-            layer.recalculate(z);
+            layer.recalculate(parameters);
             if (!layer.isHidden(z) && layer.source) {
                 this.sourceCaches[layer.source].used = true;
             }
         }
 
-        this.light.recalculate(z);
+        this.light.recalculate(parameters);
+        this.z = z;
+    }
 
-        const maxZoomTransitionDuration = 300;
-        if (Math.floor(this.z) !== Math.floor(z)) {
-            this.animationLoop.set(maxZoomTransitionDuration);
+    hasTransitions() {
+        if (this.light.hasTransition()) {
+            return true;
         }
 
-        this.z = z;
+        for (const id in this.sourceCaches) {
+            if (this.sourceCaches[id].hasTransition()) {
+                return true;
+            }
+        }
+
+        for (const id in this._layers) {
+            if (this._layers[id].hasTransition()) {
+                return true;
+            }
+        }
+
+        // if (this.placement.hasTransition()) {
```

Comment:
Need to uncomment and fix this. @ChrisLoer Is `this.placement.stillFading()` the right thing to check? It looks a bit different than [native's `hasTransitions()`](https://github.com/mapbox/mapbox-gl-native/blob/110c6a8de9942ac7090ef561ca1bed150b547150/src/mbgl/text/placement.cpp#L309-L311).

Reply:
Err.... hopefully we can improve this with a back-port from native. I think right now, the answer may be that you should ignore the placement state in this `hasTransitions`. They key logic for keeping animation running for placement/collision fading is the `needsRerender = !this.placement.isDone() || this.placement.stillFading()` part that sets `this._placementDirty` in `Map#_render`. That accounts for both collision fade animations (that's what `stillFading()` checks) but also for the map being in a state where it's waiting for the next placement to occur or in the middle of an incremental placement (kind of an analog of `stale` on the native side although it's not quite the same).

For now, if you just leave this out of `style.hasTransitions`, it should be fine because the render logic checks this state separately. /cc @ansis